app/src/main/java/fr/free/nrw/commons/wikidata/WikidataEditService.java

```diff
@@ -127,29 +135,40 @@ private void onAddCaptionResponse(Long fileEntityId, MwPostResponse response) {
     }
   }
 
-  public void createImageClaim(@Nullable final WikidataPlace wikidataPlace, final UploadResult imageUpload) {
+  public void createClaim(@Nullable final WikidataPlace wikidataPlace, final String fileName, final
+  HashMap<String, String> captions) {
     if (!(directKvStore.getBoolean("Picture_Has_Correct_Location", true))) {
-      Timber.d("Image location and nearby place location mismatched, so Wikidata item won't be edited");
+      Timber
+          .d("Image location and nearby place location mismatched, so Wikidata item won't be edited");
       return;
     }
-    editWikidataImageProperty(wikidataPlace, imageUpload);
+    addImageAndMediaLegends(wikidataPlace, fileName, captions);
   }
 
-  @SuppressLint("CheckResult")
-  private void editWikidataImageProperty(final WikidataItem wikidataItem, final UploadResult imageUpload) {
-    wikidataClient.createImageClaim(wikidataItem, String.format("\"%s\"", imageUpload.getFilename()))
-        .flatMap(revisionId -> {
-          if (revisionId != -1) {
-            return wikidataClient.addEditTag(revisionId, COMMONS_APP_TAG, COMMONS_APP_EDIT_REASON);
-          }
-          throw new RuntimeException("Unable to edit wikidata item");
-        })
-        .subscribeOn(Schedulers.io())
+  public void addImageAndMediaLegends(final WikidataItem wikidataItem, final String fileName,
+      HashMap<String, String> captions) {
+    Snak_partial p18 = new Snak_partial("value", WikidataProperties.IMAGE.getPropertyName(),
+        new ValueString(fileName.replace("File:", "")));
+
+    List<Snak_partial> snaks = new ArrayList<>();
+    for (String key : captions.keySet()) {
```

Comment:
inside the loop the value in the map is used so iterating the entryset would be better.

This could also be made to more resemble how EditClaim works where the construction of the object bears resemblance to the json but y'know optional